target/linux/ath79/dts/qca9533_tl-wr841.dtsi

```diff
@@ -0,0 +1,175 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+#include "qca9533.dtsi"
+
+/ {
+	memory@0 {
+		device_type = "memory";
+		reg = <0x0 0x2000000>;
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,115200n8";
+	};
+
+	aliases {
+		led-status = &system;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		wifi_green {
+			label = "tp-link:green:wlan";
+			gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		system: wifi_qss {
+			label = "tp-link:green:qss";
+			gpios = <&gpio 3 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		wan {
+			label = "tp-link:green:wan";
+			gpios = <&gpio 4 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		lan1 {
+			label = "tp-link:green:lan1";
+			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		lan2 {
+			label = "tp-link:green:lan2";
+			gpios = <&gpio 15 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		lan3 {
+			label = "tp-link:green:lan3";
+			gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		lan4 {
+			label = "tp-link:green:lan4";
+			gpios = <&gpio 11 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys";
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		button0 {
+			label = "Reset button";
+			linux,code = <KEY_RESTART>;
+			gpios = <&gpio 12 GPIO_ACTIVE_LOW>;
+			debounce-interval = <60>;
+		};
+
+		button1 {
+			label = "RFKILL button";
+			linux,code = <KEY_RFKILL>;
+			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
+			debounce-interval = <60>;
+		};
+	};
+};
+
+&ref {
+	clock-frequency = <25000000>;
+};
+
+&uart {
+	status = "okay";
+};
+
+&spi {
+	status = "okay";
+	num-cs = <1>;
+
+	flash@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <25000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			uboot:	partition@0 {
+				label = "u-boot";
+				reg = <0x000000 0x020000>;
+				read-only;
+			};
+
+			partition@20000 {
+				label = "firmware";
+				reg = <0x020000 0x3d0000>;
+			};
+
+			partition@3f0000 {
+				label = "art";
+				reg = <0x3f0000 0x010000>;
+				read-only;
+
+				nvmem-provider;
+				#address-cells = <1>;
+				#size-cells = <1>;
+			};
+		};
+	};
+};
+
+&mdio0 {
+	status = "okay";
+
```

Comment:
Did you take a look into the probe functions of [ag71xx_mdio.c](https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/files/drivers/net/ethernet/atheros/ag71xx/ag71xx_mdio.c) and [ag71xx_ar7240.c](https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/files/drivers/net/ethernet/atheros/ag71xx/ag71xx_ar7240.c)?

Looks like you have to define here "builtin-switch" and in phy node "phy4-mii-enable" to get it working as in ar71xx...
But i´m not sure.

Reply:
without additional changes builtin-switch is working only for mdio1

Reply:
Did you tested this?
The old code for ar71xx does the init a little bit differnet than the new of init code for ath79...
I also think the code has to be adjusted to not rely on "builtin-switch" to set "phy4-mii-enable".

Reply:
Did you remove fixed-link from eth0?
Does the eth ports work anyway?

Reply:
yes I removed it, as you can see, the PHY is `not fixed`, eth0 is not working without fixed link, there is not traffic with tcpdump on it either.
tried all the *-eth types available in the driver and with some of them, fixed link will not work either

Reply:
so I have the same problem with TL-MR3020 (i'll post the DTS, there is a small issue too :) ), but that one has `builtin-switch` on all mdio so the ethernet port is part of the switch no matter how I tried,so it struck me that I could make it act ok if I used the switch port status, like: `ucidef_set_led_switch "lan" "LAN" "tp-link:green:lan" "switch0" "0x02"`
on this device, wan port is port5 on swconfig show, I changed the led trigger like: `ucidef_set_led_switch "wan" "WAN" "tp-link:green:wan" "switch0" "0x12"` but is not working, but I'm sure that this would be right approach
any thoughts ?

Reply:
Look at page 353 in https://github.com/Deoptim/atheros/blob/master/QCA9531_V2.0_nowatermark.pdf

As u can see gmac0 could be configured to connect to the switch or directly to the phy...

Reply:
I tried it again

```
&mdio0 {
        resets = <&rst 22>;
        reset-names = "mdio";
        phy4-mii-enable;
        builtin-switch;
};

&eth0 {
        compatible = "qca,qca9530-eth", "syscon";
        resets = <&rst 9>;
        reset-names = "mac";
        pll-data = <0x82000101 0x80000101 0x80001313>;
        reg = <0x19000000 0x200
                0x18070000 0x4>;
        pll-reg = <0x4 0x2c 17>;
        pll-handle = <&pll>;
};


&mdio1 {
        resets = <&rst 23>;
        reset-names = "mdio";
        builtin-switch;
};

&eth1 {
        compatible = "qca,qca9530-eth", "syscon";
        resets = <&rst 13>;
        reset-names = "mac";
};
```

```
&mdio0 {
        status = "okay";

        phy-mask = <4>;
        phy4: ethernet-phy@4 {
                reg = <4>;
        };
};

&mdio1 {
        status = "okay";
        phy-mask = <0>;
        phy0: ethernet-phy@0 {
                reg = <0>;
        };
};

&eth0 {
        status = "okay";

        mtd-mac-address = <&uboot 0x1fc00>;
        mtd-mac-address-increment = <1>;
        phy-handle = <&phy4>;
        phy-mode = "mii";

/*      fixed-link {
                speed = <100>;
                full-duplex;
        };*/
};

&eth1 {
        status = "okay";

        mtd-mac-address = <&uboot 0x1fc00>;
        phy-handle = <&phy0>;
        phy-mode = "gmii";

/*      fixed-link {
                speed = <1000>;
                full-duplex;
        };*/
};
```
and

```
[    1.412200] libphy: mdio: probed
[    1.450459] mdio-bus.0: unknown phy id '0000:0000'
[    1.456607] ag71xx 19000000.eth: connected to PHY at mdio-bus.0:04 [uid=00000000, driver=Generic PHY]
[    1.466994] eth0: Atheros AG71xx at 0xb9000000, irq 4, mode:MII
[    2.152204] libphy: mdio: probed
[    2.195021] mdio-bus.1: Found an AR934X built-in switch
[    2.248707] ag71xx 1a000000.eth: connected to PHY at mdio-bus.1:00 [uid=004dd042, driver=Generic PHY]
[    2.259029] eth1: Atheros AG71xx at 0xba000000, irq 5, mode:GMII
```

I added phy0 to eth1 so it would start
with this setup, port4 (eth1) from the switch will show link up/down when cable is plugged, but will not work
compared to ar71xx targets, swconfig is showing additional port5 (the wan port), also on ar71xx, uid=004dd042 is detected to eth0, not eth1

setting pll to eth1 will crash the kernel if 
```
 		} else if (of_device_is_compatible(np, "qca,ar9530-eth")) {
			ath79_set_pll(ag);
```
is added to the driver


if I add phy4 to mdio1 then I have link port detect on wan port as eth1 and leds are ok, but no traffic, either to switch nor wan, but I'll try to test the options available with gmac-config

Reply:
If i´m right "phy4-mii-enable" should be set in the phy node... 
Probably you have to set "builtin-switch" and "phy4-mii-enable" in mdio1 and it´s corresponding phy node not in eth0, mdio0 nor it´s corresponding phy node...
And please set it in board dtsi and not soc dtsi.

Reply:
mdio0 doesn't exist on qca953x. We don't need this node.
PS: In the old ar71xx target the ath79_register_mdio(0, 0x0); is misleading: This function call registered mdio1 actually.

Reply:
@981213 but it is detected in dmesg, is it dummy ?